### ISSUE TITLE
Spurious deep copies in inlining

### ISSUE THREAD

**Issue by JSKenyon** [numba/numba#8989]
## Reporting a bug

- [X] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [ ] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

I recently profiled the compilation of some code in an effort be better understand where time is spent inside Numba. For my specific use case, I saw that the majority of compilation time was spent inside `deepcopy`. After some investigation, I figured out that those deep copies are associated with inlining. The code which produces these copies is here: https://github.com/numba/numba/blob/0b85096be667ed817a52f5989bf252bda552b8f1/numba/core/inline_closurecall.py#L371-L381

The problem (perhaps a bug, although I may have missed some subtlety) is that the first `deepcopy` (line 375) is followed by immediately discarding the contents of the body (line 376) and then repopulating it with additional deep copies (lines 377-379). It is this second set of copies that I am struggling to explain - surely those elements would already have been copied in the original `deepcopy` call?

The reason I am raising this is because inlining (specifically these deep copies) is currently dominating my compilation times. As an experiment, I commented out lines 376-379 and saw roughly an x2.5 speedup in compilation with no decrease in performance of the compiled code. Note that disabling inlining completely (by removing inline='always' in my code) typically results in 40% slower code in my specific use case (which is somewhat unexpected, but clearly inlining into the Numba IR helps LLVM optimize the code in some way).

To summarize: inlining produces many expensive deep copy operations - are they all necessary?

**Comment by kc611** [comment]
Thank you @JSKenyon for reporting the issue.

On closer inspection this deep-copy logic was added in https://github.com/numba/numba/pull/5812.

The problem was related to the IR being mutated in-place, a complete description for the original issue can be found at: https://github.com/numba/numba/issues/5824 and a reproducer is at: https://github.com/numba/numba/issues/5824#issuecomment-641202064

I've opened a PR #8990 to check if any known tests fails with the suggestions that you've given.

**Comment by JSKenyon** [comment]
I don't believe this issue should be marked as stale until there is an explanation of the current code.
